### ISSUE TITLE
Ложное срабатывание проверки: constructor-function-return-section

### ISSUE THREAD

**Issue by RedMammoth** [1C-Company/v8-code-style#1272]
## Название/код проверки

Код проверки: `constructor-function-return-section`

Версия плагина: `0.4.0.135`

## Пример кода/метаданных содержащих ошибку

В языке 1С:Предприятие есть объекты, которые невозможно создать в качестве значения по умолчанию в функциях-конструкторах (например, ФормаКлиентскогоПриложения, ГруппаФормы и т.п.). Поэтому в качестве значения устанавливают Неопределено, а функция использующая структуру уже заполняет конкретное значение.

В таком сценарии срабатывает проверка на поля и их остается только подавлять.

Предлагаю подумать, может можно будет такие типы анализировать в проверке и не срабатывать на них, или может быть, сделать параметр.

```bsl
// Новые параметры создания элемнтов.
// 
// Возвращаемое значение:
//  Структура - Новые параметры создания элемнтов:
// * Форма - ФормаКлиентскогоПриложения
// * Расположение - РасширениеГруппыФормыДляОбычнойГруппы
// * УникальныйПрефикс - Строка
//
Функция НовыеПараметрыСозданияЭлемнтов() Экспорт
	ПараметрыСоздания = Новый Структура;
	ПараметрыСоздания.Вставить("Форма", Неопределено);
	ПараметрыСоздания.Вставить("Расположение", Неопределено);
	ПараметрыСоздания.Вставить("УникальныйПрефикс", "");
	Возврат ПараметрыСоздания;
КонецФункции
```

![image](https://user-images.githubusercontent.com/64589646/225257118-d9efe657-e208-4956-a654-6dc9bde3a9c3.png)

**Comment by DmitryShvaika** [comment]
Аналогичная проблема с составными ссылками. Edt не знает, что значение Неопределено является умолчанием для составных ссылок

**Comment by marmyshev** [comment]
Функция возвращает `Неопределенно` - это тип, который забыли указать в типах. В чем проблема? Или функция не возвращает "неопределенно"? или хотите честно "соврать" про то что она возвращает неопределенно?

1. Либо честно указывайте типы что "форма" или "неопределенно". 
2. Либо нужно делать не "функцию-конструктор" - а "функцию-получатель" которая возвращает сразу объект с данными (т.е. с реальной формой).
3. Либо если хотите соврать про Неопределенно - надо ставить подавление.

Имхо, тут проверка как раз правильно говорит - что есть идеологическая проблема в коде - так нельзя код писать/проектировать..., "но если очень хочется - то можно" - тогда подавление. Естественно, что расставлять подавления - это неудобно, но это как раз сигнал к тому что "функции-конструкторы" новых пустых данных используются не по назначению.

**Comment by marmyshev** [comment]
> Аналогичная проблема с составными ссылками. Edt не знает, что значение Неопределено является умолчанием для составных ссылок

Не верная трактовка. 

Есть реквизиты БД составные (с любым списком типов, не обязательно ссылки) - там `неопределенно` только в том случае когда не указано пустое типизированное значение. Аналогично с ТЗ и ТД при создании колонок с указанием "Описания типа" - смысл тот же, механизм Платформы - тот же - объекты хранятся в таблицах БД в некоторых случаях, описанных на ИТС.

А в коде всё по другому, если входящий параметр составного типа - это означает что будет либо один тип, либо другой. Если в параметр приходит "Неопределенно" - нужно явно декларировать такое поведение.

**Comment by marmyshev** [comment]
Точно не баг и не ложное срабатывание. Но что делать, чтобы оставалось правильно/честно но при этом удобней чем сейчас - у меня идей нет.

Просто забить на проблему в коде (добавив отключающий параметр в проверку) - не выход, если мы планируем делать нечто аналогичное как в Эллипсе #860

**Comment by RedMammoth** [comment]
> Функция возвращает `Неопределенно` - это тип, который забыли указать в типах док.коммента. В чем проблема? Или функция не возвращает "неопределенно"? или хотите честно "соврать" про то что она возвращает неопределенно?
> 
> 1. Либо честно указывайте типы что "форма" или "неопределенно".
> 2. Либо нужно делать не "функцию-конструктор" - а "функцию-получатель" которая возвращает сразу объект с данными (т.е. с реальной формой).
> 3. Либо если хотите соврать про Неопределенно - надо ставить подавление.
> 
> Имхо, тут проверка как раз правильно говорит - что есть идеологическая проблема в коде - так нельзя код писать/проектировать..., "но если очень хочется - то можно" - тогда подавление. Естественно, что расставлять подавления - это неудобно, но это как раз сигнал к тому что "функции-конструкторы" новых пустых данных используются не по назначению.

Я зацепился за описание этой проверки https://github.com/1C-Company/v8-code-style/issues/384 , а именно за:
> При этом сама функция-конструктор не принимает параметров, а только возвращает структуру-заготовку со свойствами

Но в текущей версии описания стандарта такого требования не вижу, может проглядел (https://its.1c.ru/db/v8std#content:453:hdoc)

Если такого ограничения нет, то проблема отпадет, сделав у функции-конструктора обязательными параметрами те, которые нельзя создать и при этом должны быть заполнены, например:

```bsl
// Новые параметры создания элемнтов.
//
// Параметры:
//  Форма - ФормаКлиентскогоПриложения
//  Расположение - РасширениеГруппыФормыДляОбычнойГруппы
//
// Возвращаемое значение:
//  Структура - Новые параметры создания элемнтов:
// * Форма - ФормаКлиентскогоПриложения
// * Расположение - РасширениеГруппыФормыДляОбычнойГруппы
// * УникальныйПрефикс - Строка
//
Функция НовыеПараметрыСозданияЭлемнтов(Форма, Расположение) Экспорт
	ПараметрыСоздания = Новый Структура;
	ПараметрыСоздания.Вставить("Форма", Форма);
	ПараметрыСоздания.Вставить("Расположение", Расположение);
	ПараметрыСоздания.Вставить("УникальныйПрефикс", "");
	Возврат ПараметрыСоздания;
КонецФункции
```

Если можно так, тогда в проверке https://github.com/1C-Company/v8-code-style/issues/384 нужно допускать наличие параметров в функции-конструкторе

**Comment by marmyshev** [comment]
Да, в стандарте нет описания разницы между "функцией-конструктором" новых (и возможно, пустых) данных и "функцией-получателем" существующих данных (существующих объектов).

Имхо, функция без параметров и возвращающая сложный тип (структура, массив, ТЗ, ТД, Соответствие, Список) - это "функция-конструктор". Вроде как подходит. Всё что с параметрами - "функция-получатель".

Далее по коду, использование с именованными ссылками на возвращаемый тип - одинаковое.

**Comment by DmitryShvaika** [comment]
> Да, в стандарте нет описания разницы между "функцией-конструктором" новых (и возможно, пустых) данных и "функцией-получателем" существующих данных (существующих объектов).
> 
> Имхо, функция без параметров и возвращающая сложный тип (структура, массив, ТЗ, ТД, Соответствие, Список) - это "функция-конструктор". Вроде как подходит. Всё что с параметрами - "функция-получатель".
> 
> Далее по коду, использование с именованными ссылками на возвращаемый тип - одинаковое.

Функция-конструктор может быть и с параметрами. Например, ФиксированныйМассив или ФиксированнаяСтруктура создаются на базе простых массивов и структур. Есть функции-конструкторы для сложных структур, которые параметром принимают часть данных, например. Есть и другие примеры (например наследование).
